### PR TITLE
Remove the mutex surrounding account_bloom in state_db

### PR DESCRIPTION
**Pull request by insipx**
This is the first in a series of PR's I will make for removing some mutexes.

This is the smallest so far, so I figured it would be a good start to get thoughts on my changes.

All this one entails:

- Make `account_bloom` a struct owned by StateDB
- change `note_non_null_account` traits to a mutable reference
- add clone derivations for Bloom data structure (for cases when the DB is being `cloned`/`boxed_clone`'d)

Am I on the right track?

@sorpaas

### CODE REVIEW

File: ethcore/Cargo.toml [hunk 1]
```diff
@@ -39,7 +39,7 @@ keccak-hasher = { path = "../util/keccak-hasher" }
 kvdb = "0.1"
 kvdb-memorydb = "0.1"
 kvdb-rocksdb = { version = "0.1.3", optional = true }
-lazy_static = "1.0"
+lazy_static = "1.2.0"
```
[the remark this answers is not in the document]
Comment (ordian) [reply ethcore/Cargo.toml]: Isn't that [the same](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio)?

**Comment by ordian** [comment]
> There seem to be a very large amount of code added for what I would expect to be a very simple change. Did some commits get added that should not have been?

There was a force push to master to add coauthors to clique commit, a rebase is needed.

**Comment by insipx** [comment]
@ordian yeah, I'll get this fixer asap!

**Comment by insipx** [comment]
Fixed it, @DemiMarie-parity @ordian

**Comment by dvdplm** [comment]
I get the same 4 failing tests locally, do you know what's going on there?

**Comment by dvdplm** [comment]
@insipx what is the status of this? Are still actively working on this?

**Comment by insipx** [comment]
> @insipx what is the status of this? Are still actively working on this?

@dvdplm, I know why the issue is occurring (haven't exactly figured out a fix yet), and will be getting back to this issue soon. Progress has been slow for the past couple weeks as my Semester nears an end.

I hope to update you on this soon, however

**Comment by dvdplm** [comment]
@insipx I've looked into the test failures and they seem legit to me. I don't think this approach is going to work due to how much cloning we're doing of the `StateDb`: before this PR clones were cheap and all the clones had a ref to the same bloom data. After this PR the actual `Bloom` is cloned (so it's expensive) and things get written to the wrong copy of the `Bloom` and later mistakenly assumed to not be there.
I'm closing this now as the fix seem non-trivial and not actively being worked on. Please feel free to re-open at any time or contact me to discuss further.

**Comment by dvdplm** [comment]
My investigation branch is [here](https://github.com/paritytech/parity-ethereum/tree/dp/insipx/insi_mutex-account-bloom) if we need it.

**Pull request closed without merging**